src/coprocessor/dag/aggr_fn/impl_count.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+use cop_datatype::{FieldTypeFlag, FieldTypeTp};
+use tipb::expression::{Expr, ExprType, FieldType};
+
+use crate::coprocessor::codec::data_type::*;
+use crate::coprocessor::codec::mysql::Tz;
+use crate::coprocessor::dag::expr::EvalContext;
+use crate::coprocessor::dag::rpn_expr::{RpnExpression, RpnExpressionBuilder};
+use crate::coprocessor::Result;
+
+/// The parser for COUNT aggregate function.
+pub struct AggrFnDefinitionParserCount;
+
+impl super::parser::Parser for AggrFnDefinitionParserCount {
+    fn check_supported(&self, aggr_def: &Expr) -> Result<()> {
+        assert_eq!(aggr_def.get_tp(), ExprType::Count);
+        if aggr_def.get_children().len() != 1 {
+            return Err(box_err!(
+                "Expect 1 parameter, but got {}",
+                aggr_def.get_children().len()
+            ));
+        }
+
+        // Only check whether or not the children expr is supported.
+        let child = &aggr_def.get_children()[0];
+        RpnExpressionBuilder::check_expr_tree_supported(child)?;
+
+        Ok(())
+    }
+
+    fn parse(
+        &self,
+        mut aggr_def: Expr,
+        time_zone: &Tz,
+        max_columns: usize,
+        out_schema: &mut Vec<FieldType>,
+        out_exp: &mut Vec<RpnExpression>,
+    ) -> Result<Box<dyn super::AggrFunction>> {
+        use cop_datatype::FieldTypeAccessor;
+
+        assert_eq!(aggr_def.get_tp(), ExprType::Count);
+        let child = aggr_def.take_children().into_iter().next().unwrap();
+
+        // COUNT outputs one column.
+        out_schema.push({
+            let mut ft = FieldType::new();
+            ft.as_mut_accessor()
+                .set_tp(FieldTypeTp::LongLong)
+                .set_flag(FieldTypeFlag::UNSIGNED);
+            ft
+        });
+
+        // COUNT doesn't need to cast, so using the expression directly.
+        out_exp.push(RpnExpressionBuilder::build_from_expr_tree(
+            child,
+            time_zone,
+            max_columns,
+        )?);
+
+        Ok(Box::new(AggrFnCount))
+    }
+}
+
+/// The COUNT aggregate function.
+#[derive(Debug)]
+pub struct AggrFnCount;
+
+impl super::AggrFunction for AggrFnCount {
+    #[inline]
+    fn name(&self) -> &'static str {
+        "AggrFnCount"
+    }
+
+    #[inline]
+    fn create_state(&self) -> Box<dyn super::AggrFunctionState> {
```

Comment:
Could we merge AggrFnStateCount and AggrFnCount  to one? Too many structs make it very hard to review

Reply:
AggrFnState and AggrFn are just different things. For different groups we use the same `AggrFn` instance but different `AggrFnState`s.